proposals/clone_mulitple_resource.md

```diff
@@ -0,0 +1,173 @@
+# Meta
+[meta]: #meta
+- Name: Allow cloning multiple resources from a Namespace
+- Start Date: 2022-08-04
+- Update data (optional): 2022-08-04
+- Author: [Prateek Pandey](https://github.com/prateekpandey14)
+- Supersedes:  N/A
+
+# Table of Contents
+[table-of-contents]: #table-of-contents
+- [Meta](#meta)
+- [Table of Contents](#table-of-contents)
+- [Overview](#overview)
+- [Definitions](#definitions)
+- [Motivation](#motivation)
+- [Proposal](#proposal)
+- [Implementation](#implementation)
+- [Migration (OPTIONAL)](#migration-optional)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Prior Art](#prior-art)
+- [Unresolved Questions](#unresolved-questions)
+- [CRD Changes (OPTIONAL)](#crd-changes-optional)
+
+# Overview
+[overview]: #overview
+
+Enhance kyverno generate rule to clone mulitple resources from a given namespace.
+
+
+# Definitions
+[definitions]: #definitions
+
+N/A
+
+# Motivation
+[motivation]: #motivation
+
+The initial idea for this feature came from issue [282](https://github.com/kyverno/kyverno/issues/282) and received feedback from the kyverno community in slack.
+
+It would be good to have a way to copy/clone multiple resources from a namespace in a single rule, instead of
+writing a new rule or policy.
+
+There are primarily main use cases associated with this.
+
+1. To clone set of resources from a common namespace, using a match selector whenever we create a new namespace
+   to bootstrap the deployment process.
+
+2. Implicit, User has to less number of policy/rule to copy the set of necessary resources.
+
+# Proposal
+
+In this proposal, Kyverno generate rule can be extended to allowing to clone multiple resources from a namespace.
+This could be done using  matching selector to select multiple resources of the given apiVersion/group and allowing
+multiple kinds/name pairs, similar to the match/exclude blocks we use to filter resources in the admission control
+webhooks.
+
+In below extended ClusterPolicy example, multiple resources can be selected using a match label selector `rollout: true`
+to copy/clone from `bootstrap` namespace to newly created namespace.
+
+```yaml
+    generate:
+      namespace: "{{request.object.metadata.name}}"
+      synchronize : true
+      clone:
+        namespace: bootstrap    // common namespace
+        kinds:
+          - v1/Secret
+        selector:
+          matchLabels:
+            rollout: "true"    // clone resource with matching selector i.e. configmap, secrets etc
+```
+
+### Examples:
+
+1. Clone multiple resource using matching kinds list
+
+```yaml
+
+apiVersion: kyverno.io/v1
+kind: ClusterPolicy
+metadata:
+  name: clone-resource-with-selector
+spec:
+  generateExistingOnPolicyUpdate: true
+  rules:
+  - name: clone-resource
+    match:
+      any:
+      - resources:
+          kinds:
+          - Namespace
+    generate:
+      namespace: "{{request.object.metadata.name}}"
+      synchronize : true
+      cloneList:
+        - namespace:
```

Comment:
```suggestion
        - namespace: staging
```

As discussed in the sync, this should not be a list. Because each entry in the cloneList only needs a single namespace field.

Reply:
Updated